internals/daemon/daemon.go

```diff
@@ -142,8 +142,22 @@ const (
 	accessForbidden
 )
 
-func userFromRequest(state interface{}, r *http.Request) (*UserState, error) {
-	return nil, nil
+func userFromRequest(st *state.State, r *http.Request, ucred *Ucrednet) (*UserState, error) {
+	if ucred == nil {
+		// No ucred details, no UserState. Currently, "local" (ucred-based) is
+		// the only type of identity we support.
+		return nil, nil
+	}
+
+	st.Lock()
+	identity := st.IdentityFromInputs(&ucred.Uid)
+	st.Unlock()
+
+	if identity == nil {
+		// No identity that matches these inputs (for now, just UID).
+		return nil, nil
+	}
+	return &UserState{Access: identity.Access, UID: &ucred.Uid}, nil
```

Comment:
It looks like (I didn't look too closesly) that `IdentityFromInputs()` may return a pointer to an object inside the identities structure (which is part of `st`, and protected by the mutex?), and we later access `identity.Access` in line 160 without the lock held.

Better to keep the lock until it's no longer possible to access `identity` (which currently is at the end of the function). Checking for `identity == nil` and creating a `UserState` struct shouldn't take up too much time, so the lock shouldn't be kept held unnecessarily long this way.

```suggestion
	st.Lock()
	identity := st.IdentityFromInputs(&ucred.Uid)
	defer st.Unlock()

	if identity == nil {
		// No identity that matches these inputs (for now, just UID).
		return nil, nil
	}
	return &UserState{Access: identity.Access, UID: &ucred.Uid}, nil
```

Reply:
I believe the trick is that identity objects never get mutated (they get replaced), so once you have a pointer to an instance, its safe to keep it.

Reply:
Yep, Fred's right here -- thanks.